### ISSUE TITLE
Disconnect trigger is not invoked when a socket is closed with outstanding input

### ISSUE THREAD

**Issue by kostja** [tarantool/tarantool#2998]
This may be a security issue as long as we begin using  on_disconnect triggers for audit:
```static inline void                                                              
iproto_connection_close(struct iproto_connection *con)        
        /*                                                                      
         * If the connection has no outstanding requests in the                 
         * input buffer, then no one (e.g. tx thread) is referring              
         * to it, so it must be destroyed at once. Queue a msg to               
         * run on_disconnect() trigger and destroy the connection.              
         *                                                                      
         * Otherwise, it will be destroyed by the last request on               
         * this connection that has finished processing.                        
         *                                                                      
         * The check is mandatory to not destroy a connection                   
         * twice.                                                               
         */                                                                     
        if (iproto_connection_is_idle(con)) {                                   
                assert(con->disconnect != NULL);                                
                struct iproto_msg *msg = con->disconnect;                       
                con->disconnect = NULL;                                         
                cpipe_push(&tx_pipe, msg);                                      
        }     
```

             
Suggested fix: always delete the connection by pushing con->disconnect message through tx pipe.

**Comment by Gerold103** [comment]
It is called, when outstanding input is processed. Test case:
```Lua
-- Server --
box.cfg{listen=3313}
box.schema.user.grant('guest', 'read,write,execute', 'universe')
disconnected = false
long_is_in_progress = false
long_fiber = nil
fiber = require('fiber')
function set_on_disconnect()
	box.session.on_disconnect(function() disconnected = true end)
end
function long_poll()
	long_fiber = fiber.self()
	long_is_in_progress = true
	fiber.sleep(3000)
end

-- Client --
netbox = require('net.box')
fiber = require('fiber')
c = netbox:connect(3313)
c:call('set_on_disconnect')
fiber.create(function() c:call('long_poll') end)
```

Then close connection from client:
```Lua
c:close()
```

On the server `iproto_connection_close()` is called. Then cancel the long fiber (which makes connection in not idle state).
```Lua
long_fiber:cancel()
```

Right after that the `on_disconnect` is called, and `disconnected` flag is set to true.

**Comment by kostja** [comment]
Please add a test case and close the bug.

**Comment by Gerold103** [comment]
Closed in 32f1c3489990
